verifier/src/x509/chain.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2023 The MobileCoin Foundation
+
+//! Support for verifying certificate chains
+
+extern crate alloc;
+
+use super::certs::UnverifiedCertificate;
+use super::Result;
+use crate::x509::certs::VerifiedCertificate;
+use alloc::vec::Vec;
+use core::time::Duration;
+use p256::ecdsa::VerifyingKey;
+
+/// An X509 certificate chain. This is a valid path from the trust root to the
+/// leaf certificate.
+pub struct CertificateChain {
+    certificates: Vec<UnverifiedCertificate>,
+}
+
+impl CertificateChain {
+    /// Create a new certificate chain from a path from a trust root to the
+    /// leaf certificate.
+    ///
+    /// A certificate chain without a valid path will result in errors for
+    /// functions like [`CertificateChain::signing_key`].
+    pub fn new(certificates: Vec<UnverifiedCertificate>) -> Self {
+        Self { certificates }
+    }
+
+    /// Returning the signing key from the leaf certificate
+    ///
+    /// The chain will be verified against the `trust_root` and the `unix_time`.
+    pub fn signing_key(
+        &self,
+        trust_root: VerifiedCertificate,
+        unix_time: Duration,
+    ) -> Result<VerifyingKey> {
```

Comment:
This might be doing too much, but TBH I'm still on the fence on the "unverified***"->"verified***" logic I have for certs and CRLs.